### PR TITLE
MAN-26: Added caseload search filter and sort

### PR DESCRIPTION
**Pull request by pmcphee77**

### CODE REVIEW

File: projects/manage-supervision-and-delius/src/main/kotlin/uk/gov/justice/digital/hmpps/integrations/delius/user/entity/User.kt [hunk 1]
```diff
@@ -218,20 +223,194 @@ data class Caseload(
...
+        join fetch c.team t
+        left join fetch c.nextAppointment na
+        left join fetch na.type naType
+        left join fetch c.previousAppointment pa
+        left join fetch pa.type paType
+        left join fetch c.latestSentence ls
+        left join fetch ls.mainOffence mo
+        left join fetch mo.offence moo
+        where c.staff.code = :staffCode
+        and ((:nameOrCrn is null or upper(p.crn) like '%' || upper(:nameOrCrn) || '%' or upper(p.forename || ' ' || p.surname) like '%' || upper(:nameOrCrn) || '%'))
```
Comment (marcus-bcl) [inline projects/manage-supervision-and-delius/src/main/kotlin/uk/gov/justice/digital/hmpps/integrations/delius/user/entity/User.kt:401]: Should this include a filter on surname+forename too?  E.g. if someone searched for "Smith, John"?
```suggestion
        and (
          :nameOrCrn is null 
          or upper(p.crn) like '%' || upper(:nameOrCrn) || '%' 
          or upper(p.forename || ' ' || p.surname) like '%' || upper(:nameOrCrn) || '%'
          or upper(p.surname || ' ' || p.forename) like '%' || upper(:nameOrCrn) || '%'
          or upper(p.surname || ', ' || p.forename) like '%' || upper(:nameOrCrn) || '%')
```

Feel free to ignore this if you think it makes the query too verbose though.

File: projects/manage-supervision-and-delius/src/main/kotlin/uk/gov/justice/digital/hmpps/api/controller/CaseloadController.kt [hunk 2]
```diff
@@ -37,4 +51,17 @@ class CaseloadController(private val userService: UserService) {
     @GetMapping("/team/{teamCode}/staff")
     @Operation(summary = "Gets the staff within the team")
     fun getTeamStaff(@PathVariable teamCode: String) = userService.getTeamStaff(teamCode)
+
+    private fun sort(sortString: String): Sort {
+
+        val regex = Regex(pattern = "[A-Za-z0-9]+\\.(ASC|DESC)", options = setOf(RegexOption.IGNORE_CASE))
```
Comment (marcus-bcl) [inline projects/manage-supervision-and-delius/src/main/kotlin/uk/gov/justice/digital/hmpps/api/controller/CaseloadController.kt:57]: nitpick (feel free to ignore) - do you need A-Za-z when the regex is case-insensitive?
```suggestion
        val regex = Regex(pattern = "[A-Z0-9]+\\.(ASC|DESC)", options = setOf(RegexOption.IGNORE_CASE))
```

File: projects/manage-supervision-and-delius/src/main/kotlin/uk/gov/justice/digital/hmpps/integrations/delius/user/entity/User.kt [hunk 3]
```diff
@@ -218,20 +223,194 @@ data class Caseload(
...
+            select e.*,
+            row_number() over (partition by e.offender_id order by cast(e.event_number as number) desc) as row_num 
+            from event e join main_offence mo on mo.event_id = e.event_id
+            where e.soft_deleted = 0 
+            and e.active_flag = 1
+            ) sub
+        where sub.row_num = 1
+"""
+)
+@Synchronize("event")
```
Comment (marcus-bcl) [inline projects/manage-supervision-and-delius/src/main/kotlin/uk/gov/justice/digital/hmpps/integrations/delius/user/entity/User.kt:261]: Nice - I've not really used `@Subselect` + `@Synchronize` before. Looks clever 😄

File: projects/manage-supervision-and-delius/src/main/kotlin/uk/gov/justice/digital/hmpps/integrations/delius/user/entity/User.kt [hunk 4]
```diff
@@ -218,20 +223,194 @@ data class Caseload(
...
+@Subselect(
+    """
+        select * from (
+            select 
+            q.staff_id,
+            q.offender_id,
+            q.appointment_datetime,
+            q.contact_id,
+            q.contact_type_id,
+            row_number() over (partition by staff_id, offender_id order by appointment_datetime asc) as row_num
```
Comment (marcus-bcl) [inline projects/manage-supervision-and-delius/src/main/kotlin/uk/gov/justice/digital/hmpps/integrations/delius/user/entity/User.kt:283]: Why do we need to partition by staff id here?  Are we only showing the latest/next appointment for the current user with the person on probation?

**Review by marcus-bcl — COMMENTED** [review]
Looks good!  Just added a couple of questions

File: projects/manage-supervision-and-delius/src/main/kotlin/uk/gov/justice/digital/hmpps/integrations/delius/user/entity/User.kt (hunk 4 above)
Comment (pmcphee77) [reply projects/manage-supervision-and-delius/src/main/kotlin/uk/gov/justice/digital/hmpps/integrations/delius/user/entity/User.kt:283]: we are actually looking for the latest appointment for the staff record here as we are looking for the staffs appointment with the crn. The appointment for the crn could in theory be with another staff member

File: projects/manage-supervision-and-delius/src/main/kotlin/uk/gov/justice/digital/hmpps/integrations/delius/user/entity/User.kt (hunk 1 above)
Comment (pmcphee77) [reply projects/manage-supervision-and-delius/src/main/kotlin/uk/gov/justice/digital/hmpps/integrations/delius/user/entity/User.kt]: agreed! i like it. Thanks. Done

File: projects/manage-supervision-and-delius/src/main/kotlin/uk/gov/justice/digital/hmpps/api/controller/CaseloadController.kt (hunk 2 above)
Comment (pmcphee77) [reply projects/manage-supervision-and-delius/src/main/kotlin/uk/gov/justice/digital/hmpps/api/controller/CaseloadController.kt]: Yep - dont need that. Changed

File: projects/manage-supervision-and-delius/src/main/kotlin/uk/gov/justice/digital/hmpps/integrations/delius/user/entity/User.kt (hunk 3 above)
Comment (pmcphee77) [reply projects/manage-supervision-and-delius/src/main/kotlin/uk/gov/justice/digital/hmpps/integrations/delius/user/entity/User.kt:261]: Yeah handy for these situations

**Review by marcus-bcl — APPROVED** [review]

**Pull request merged**